vtStorAtaProtocol/Platform/Linux/ProtocolAtaPassThrough.cpp

```diff
@@ -27,164 +31,141 @@ namespace vtStor
-        m_AtaPassThrough.CurrentTaskFile[COUNT_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Count;
-        m_AtaPassThrough.CurrentTaskFile[LBA_LOW_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.LbaLow;
-        m_AtaPassThrough.CurrentTaskFile[LBA_MID_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.LbaMid;
-        m_AtaPassThrough.CurrentTaskFile[LBA_HIGH_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.LbaHigh;
-        m_AtaPassThrough.CurrentTaskFile[DEVICE_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Device;
-        m_AtaPassThrough.CurrentTaskFile[COMMAND_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Command;
-        m_AtaPassThrough.CurrentTaskFile[RESERVED_REGISTER_OFFSET] = CurrentTaskFile.InputRegister.Reserved;
+        m_Cdb16.ExtendProtocol |= SG_ATA_LBA48;
     }
 
-    eErrorCode cAtaPassThrough::IssuePassThroughDirectCommand( const DeviceHandle& Handle, U32& BytesReturned )
+    //! If transfer mode is DMA
+    if (StorageUtility::Ata::eTransferMode::DMA_PROTOCOL == AtaCommandCharacteristic.TransferMode)
+    {
+        m_Cdb16.ExtendProtocol |= SG_ATA_PROTO_DMA;
+    }
+
+    m_Cdb16.Param |= (SG_CDB2_TLEN_NSECT | SG_CDB2_TLEN_SECTORS);
+
+    //! Set data access flag
+    switch (AtaCommandCharacteristic.DataAccess)
+    {
+    case StorageUtility::Ata::eDataAccess::READ_FROM_DEVICE:
+    {
+        m_ATAPassThrough.dxfer_direction = SG_DXFER_FROM_DEV;
+        m_Cdb16.ExtendProtocol = SG_ATA_PROTO_PIO_IN;
+        m_Cdb16.Param |= SG_CDB2_TDIR_FROM_DEV;
+    } break;
+
+    case StorageUtility::Ata::eDataAccess::WRITE_TO_DEVICE:
     {
-        
-        assert( INVALID_HANDLE_VALUE != Handle );
-
-        eErrorCode error;
-        error = eErrorCode::None;
-        
-        BOOL commandSuccessfulFlag;
-        DWORD bytesReturned;
-        commandSuccessfulFlag = DeviceIoControl
-            (
-            Handle,
-            IOCTL_ATA_PASS_THROUGH_DIRECT,
-            &m_AtaPassThrough,
-            m_AtaPassThrough.Length,
-            &m_AtaPassThrough,
-            m_AtaPassThrough.Length,
-            &bytesReturned,
-            NULL
-            );
-
-        // If the operation was not successful:
-        if (FALSE == commandSuccessfulFlag)
-        {
-            error = eErrorCode::Io;
-            
-            //TODO: report extended error
-            //fprintf( stderr, "\nDeviceIoControl was not successful. Error Code: %d", GetLastError() );
-        }
-
-        BytesReturned = bytesReturned;
-
-        return(error);
-    }*/
+        m_ATAPassThrough.dxfer_direction = SG_DXFER_TO_DEV;
+        m_Cdb16.ExtendProtocol = SG_ATA_PROTO_PIO_OUT;
+        m_Cdb16.Param |= SG_CDB2_TDIR_TO_DEV;
+    } break;
+
+    case StorageUtility::Ata::eDataAccess::NONE:
+    {
+        m_ATAPassThrough.dxfer_direction = SG_DXFER_NONE;
+        m_ATAPassThrough.dxfer_len = 0;
+        m_Cdb16.ExtendProtocol = SG_ATA_PROTO_NON_DATA;
+        m_Cdb16.Param = SG_CDB2_CHECK_COND;
+    } break;
+

```

Comment:
enum class eDataAccess
{
    NONE,
    READ_FROM_DEVICE,
    WRITE_TO_DEVICE
};

There are 3 cases, but i will add default case to catch cases which add in other version